lib/rspec/core/source.rb

```diff
@@ -0,0 +1,71 @@
+RSpec::Support.require_rspec_core 'source/node'
+RSpec::Support.require_rspec_core 'source/token'
+
+module RSpec
+  module Core
+    # @private
+    # Represents a Ruby source file and provides access to AST and tokens.
+    class Source
+      attr_reader :source, :path
+
+      def self.from_file(path)
+        source = File.read(path)
+        new(source, path)
+      end
+
+      def initialize(source_string, path=nil)
+        @source = source_string
+        @path = path ? File.expand_path(path) : '(string)'
+      end
+
+      def lines
+        @lines ||= source.split("\n")
+      end
+
+      def ast
+        @ast ||= begin
+          require 'ripper'
```

Comment:
> I thought this file is only loaded if it's being used to extract a snippet, and that can't be done w/o ripper, right?

`Source` is used even when extracting a single line in Ripper unsupported environment.

https://github.com/rspec/rspec-core/blob/f8d6820486730970afc73399c967c143895a936a/lib/rspec/core/formatters/snippet_extractor.rb#L11-L16